lib/pbench/server/api/resources/query_apis/metadata_index/iteration_samples.py

```diff
@@ -0,0 +1,356 @@
+from http import HTTPStatus
+from logging import Logger
+from typing import AnyStr, List, Union
+
+from flask import Response, jsonify
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    JSON,
+    Parameter,
+    ParamType,
+    Schema,
+)
+from pbench.server.api.resources.query_apis import CONTEXT, PostprocessError
+from pbench.server.api.resources.query_apis.metadata_index import RunIdBase
+from pbench.server.database.models.template import Template, TemplateNotFound
+
+
+class IterationSampleNamespace(RunIdBase):
+    """
+    Iteration samples API that returns namespace for each whitelisted
+    result-data-sample subdocuments as well as lists of available values for
+    each name.
+    """
+
+    WHITELIST_AGGS_FIELDS = ["run", "sample", "iteration", "benchmark"]
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(Parameter("run_id", ParamType.STRING, required=True)),
+        )
+
+    def get_keyword_fields(
+        self, mappings: JSON, prefix: AnyStr = "", result: Union[List, None] = None
+    ) -> List:
+        if result is None:
+            result = []
+        if "properties" in mappings:
+            for p, m in mappings["properties"].items():
+                self.get_keyword_fields(m, f"{prefix}{p}.", result)
+        elif mappings.get("type") == "keyword":
+            result.append(prefix[:-1])  # Remove the trailing dot, if any
+        else:
+            for f, v in mappings.get("fields", {}).items():
+                self.get_keyword_fields(v, f"{prefix}{f}.", result)
+        return result
+
+    def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
+        """
+        Construct a pbench Elasticsearch query for aggregating a list of
+        values for keyword fields mentioned in the WHITELIST_AGGS_FIELDS
+        sub-documents of result-data-sample index document that belong to the
+        given run id.
+
+        Args:
+            json_data: JSON dictionary of type-normalized parameters
+                "run_id": Dataset document ID
+
+        EXAMPLE:
+        {
+            "run_id": "1234567890"
+        }
+        """
+        run_id = context["run_id"]
+        dataset = context["dataset"]
+
+        self.logger.info(
+            "Return iteration sample namespace for dataset {}, prefix {}, " "run id {}",
+            dataset,
+            self.prefix,
+            run_id,
+        )
+
+        # Retrieve the ES indices that belong to this run_id from the metadata
+        # table
+        index = self.get_index(dataset, "result-data-sample")
```

Comment:
Can we rename this to `indices` to indicate the plural nature of the value?

Reply:
Fixed

Reply:
Ah I missed it in a `namespace` class

Reply:
now fixed